src/lib.rs

```diff
@@ -541,55 +552,139 @@ where
             result?;
         }
 
-        self.writer
-            .write_block(bounds.clone(), items.len(), &self.compress_buffer)
+        let cur_offset = self.cursor;
+        let reg = ShardRecord {
+            offset: self.cursor,
+            start_key: bounds.0,
+            end_key: bounds.1,
+            len_bytes: self.compress_buffer.len(),
+            len_items: items.len(),
+        };
+
+        self.regions.push(reg);
+        self.cursor += self.compress_buffer.len();
+        let l = self
+            .file
+            .write_at(&self.compress_buffer, cur_offset as u64)?;
+
+        Ok(l)
     }
 
     /// Write out the shard positioning data
-    fn write_index_block(&mut self) -> Result<(), Error> {
+    fn write_index(&mut self) -> Result<(), Error> {
         let mut buf = Vec::new();
 
-        serialize_into(&mut buf, &self.writer.regions)?;
+        serialize_into(&mut buf, &self.regions)?;
 
-        let index_block_position = self.writer.cursor;
+        let index_block_position = self.cursor;
         let index_block_size = buf.len();
 
-        self.writer
-            .file
+        self.file
             .write_at(buf.as_slice(), index_block_position as u64)?;
 
-        self.writer.file.seek(SeekFrom::Start(
+        self.file.seek(SeekFrom::Start(
             (index_block_position + index_block_size) as u64,
         ))?;
-        self.writer.file.write_u64::<BigEndian>(0 as u64)?;
-        self.writer
-            .file
+        self.file.write_u64::<BigEndian>(0 as u64)?;
+        self.file
             .write_u64::<BigEndian>(index_block_position as u64)?;
-        self.writer
-            .file
-            .write_u64::<BigEndian>(index_block_size as u64)?;
+        self.file.write_u64::<BigEndian>(index_block_size as u64)?;
         Ok(())
     }
 }
 
+/// Sort buffered items, break large buffer into chunks.
+/// Serialize, compress and write the each chunk. The
+/// file manager maintains the index data.
+struct ShardWriterInner<T, S>
+where
+    T: Send + Serialize,
+    S: SortKey<T>,
+    <S as SortKey<T>>::Key: Ord + Clone + Serialize,
+{
+    sender_buffer_size: usize,
+    state_machine: BufferStateMachine<T, SortAndWriteHandler<T, S>>,
+    closed: AtomicBool,
```

Comment:
Doesn't look like this AtomicBool is set to true anywhere. Is it needed in addition to the `closed` in `BufferStateMachine`?

Reply:
removed.